src/Revisionable/Document/Repository/RevisionRepository.php

```diff
@@ -0,0 +1,182 @@
+     *
+     * Callers to this method will need to persist and flush changes to the document.
+     *
+     * @param T            $document
+     * @param positive-int $version
+     *
+     * @throws UnexpectedValueException
+     */
+    public function revert(object $document, int $version = 1): void
+    {
+        $documentWrapper = new MongoDocumentWrapper($document, $this->getDocumentManager());
+
+        $documentMetadata = $documentWrapper->getMetadata();
+        $documentId = (string) $documentWrapper->getIdentifier(false, true);
+        $documentClass = $documentMetadata->getName();
+
+        $qb = $this->createQueryBuilder();
+        $qb->field('revisionableId')->equals($documentId);
+        $qb->field('revisionableClass')->equals($documentClass);
+        $qb->field('version')->lte($version);
+        $qb->sort('version', 'ASC');
+
+        $revisions = $qb->getQuery()->getIterator()->toArray();
+
+        if ([] === $revisions) {
+            throw new UnexpectedValueException(sprintf('Could not find any revisions for version %d of document %s.', $version, $documentClass));
+        }
+
+        $data = [[]];
+
+        while ($revision = array_shift($revisions)) {
+            $data[] = $revision->getData();
+        }
+
+        $data = array_merge(...$data);
+
+        $this->fillDocument($document, $data);
+    }
+
+    /**
+     * Fills a document's versioned fields with the given data
+     *
+     * @param T                    $document
+     * @param array<string, mixed> $data
+     */
+    protected function fillDocument(object $document, array $data): void
+    {
+        $documentWrapper = new MongoDocumentWrapper($document, $this->getDocumentManager());
+
+        $documentMeta = $documentWrapper->getMetadata();
+
+        assert($documentMeta instanceof ClassMetadata);
+
+        $config = $this->getListener()->getConfiguration($this->getDocumentManager(), $documentMeta->getName());
+        $fields = $config['versioned'];
+
+        foreach ($data as $field => $value) {
+            if (!in_array($field, $fields, true)) {
+                continue;
+            }
+
+            $mapping = $documentMeta->getFieldMapping($field);
+
+            // Fill the embedded document
+            if ($documentWrapper->isEmbeddedAssociation($field)) {
+                if (!empty($value)) {
+                    assert(class_exists($mapping['targetDocument']));
+
+                    $embeddedMetadata = $this->getDocumentManager()->getClassMetadata($mapping['targetDocument']);
+                    $document = $embeddedMetadata->newInstance();
+                    $this->fillDocument($document, $value);
+                    $value = $document;
+                }
+            } elseif ($documentMeta->isSingleValuedAssociation($field)) {
+                assert(class_exists($mapping['targetDocument']));
+
+                $value = $value ? $this->getDocumentManager()->getReference($mapping['targetDocument'], $value) : null;
```

Comment:
Could we use a stricter condition around `$value`?

Reply:
Maybe a `!== null` would work in this branch?  It'd be another one of those "we gotta test it thoroughly" things, and TBH there's only a very basic "does it work without erroring" smoke test for the `revert()` capabilities so we'd need to build out some good test fixtures here.

Reply:
This is where me not knowing the MongoDB ODM well doesn't help any.

```
1 "isSingleValuedAssociation"
2 "67ce41458a6433683e0c45e6"
```

This is the only time the branch gets reached in the `RevisionableDocumentTest`.  Looking at the comment fixture, it looks like this is going to be the `ReferenceOne` relation to the article, so I guess `$value` in this case is the document identifier?  If that's the case then this could most likely work as a `!== null` check, assuming that it goes from a non-null to a null value when a non-required reference is removed.

Reply:
Friendly ping @franmomu :pray:

Based on your experience, maybe you have a chance to help with this.